### PR TITLE
Complete revise

### PR DESCRIPTION
**Pull request by nagmat84**
**Main achievements**

- The callback which is called when the dialog becomes ready receives a dictionary of (name, html element)-pairs for all HTML form elements. This way the calling code can add event listeners to theses elements without using CSS selectors.
- The callbacks which are called when the dialog is closed or cancelled receive a dictionary of (name, values)-pairs for HTML input, select and textarea elements with proper converted types, i.e. checkboxes return proper booleans, number fields numbers, etc.
- Support for textareas added

**Minor achievements**

- Added a lot of jsDoc type annotation
- No repeated, unnecessary and inefficient CSS selectors anymore; HTML elements which have once been queried for are cached

**Problems**

I was not able to build the project. When I run `npm install` to install the necessary dependencies, `npm install` breaks with

    ../src/create_string.cpp:17:37: Error: no matching function signature for calling v8::String::Utf8Value::Utf8Value(v8::Local<v8::Value>&)«

 while some tool named `gyp` tries to build the package `sass/node-sass`. Probably, the dependencies are completely ancient and not compatible with modern Node.js. I have not yet looked into updating the dependencies. If someone finds the time for that, this would be awesome.

### CODE REVIEW

**Review by ildyria — APPROVED** [review]

**Comment by ildyria** [comment]
Btw this basic Modal will be nuked once Livewire front-end is done :)

File: src/scripts/main.js [hunk 1]
```diff
@@ -1,366 +1,570 @@
...
+	 * The HTML element which had the focus before the modal dialog became visible.
+	 *
+	 * @type {?HTMLElement}
+	 */
+	lastFocus: null,
+
+	/**
+	 * Holds the container of a Basic Modal Dialog, if it is visible.
+	 *
+	 * The outer structure of a Basic Model consists of three nested DIVs:
```
Comment (kamil4) [inline src/scripts/main.js]:
```suggestion
	 * The outer structure of a Basic Modal consists of three nested DIVs:
```

File: src/scripts/main.js [hunk 2]
```diff
@@ -1,366 +1,570 @@
...
+	_dom.dialog.addEventListener('input', _removeErrorClass);
+	_dom.dialog.addEventListener('blur', _removeErrorClass);
+	_dom.dialog.addEventListener('change', _removeErrorClass);
 }
 
-export const show = function(data) {
+/**
+ * Constructs and shows a modal dialog.
+ *
+ * After the dialog has become ready, the callback `data.callback` is
```
Comment (kamil4) [inline src/scripts/main.js]:
```suggestion
 * After the dialog has become ready, the callback `confData.readyCB` is
```

**Review by kamil4 — APPROVED** [review]
Holy cow that thing must've bothered you! I thought you would simply add support for textareas but this is pretty much a rewrite...

**Comment by nagmat84** [comment]
Yes, originally I didn't remember you demand for textareas, but I was annoyed by two problems in the frontend:

a) the action handler received result data which was unusable in most of the cases, in particular for checkboxes, because the values of checkboxes were reported wrongly;

b) we had to hack into the method `show` to install some additional event handlers on input elements the allow correct interactions of input elements (`setPublic` in the frontend is a prime example for that).

So I decided to see if I can do something about it. Only later I remembered that you needed textareas, too.

**Comment by ildyria** [comment]
That's quite a complete rewrite, also am I wrong or you tried to nuke jQuery deps ? :)
Code LGTM.

**Comment by nagmat84** [comment]
Officially, there has never been a dependency on jQuery (see `README.md` and package dependencies in `package.json`). I believe the only jQuery selectors were added to the code at some later point.

However, queriing the DOM over and over again for a HTMLNode you already have is not advised. (My IDE also yells at me for that in the frontend code).

Moreover, nowadays with the standardization of the Web APIs jQuery mostly lost its primary purpose. jQuery had its right for existence when each browser had its own proprietary DOM, but these times are gone.

**Comment by ildyria** [comment]
> Moreover, nowadays with the standardization of the Web APIs jQuery mostly lost its primary purpose. jQuery had its right for existence when each browser had its own proprietary DOM, but these times are gone.

Syntactic sugar & conciseness ? :)

**Comment by nagmat84** [comment]
I would like to leave this PR open until I adopted the Lychee frontend to it. Maybe, this will unveil one or another bug. I see no other chance to test this PR.

**Pull request merged**